script/model/forecast.py

```diff
@@ -0,0 +1,585 @@
+        :param test_seq:  1D Test sequence
+        :return: Predicted value at certain horizon
+        """
+        pass
+
+
+class ForecastTrainer:
+    """
+    A wrapper around various ForecastModels, that prepares training and evaluation data.
+    """
+
+    def __init__(
+            self,
+            data: np.ndarray,
+            test_mode: bool = False,
+            eval_size: int = EVAL_DATA_SIZE,
+            seq_len: int = SEQ_LEN,
+            horizon_len: int = HORIZON_LEN) -> None:
+        """
+        :param data:  1D time-series data for a query cluster
+        :param test_mode: True If the Loader is for testing
+        :param eval_size: Number of data points used for evaluation(testing)
+        :param seq_len: Length of a sequence
+        :param horizon_len: Horizon length
+        """
+        self.seq_len = seq_len
+        self.horizon_len = horizon_len
+        self.test_mode = test_mode
+        self.eval_data_size = eval_size
+
+        self.train_raw_data, self.test_raw_data = self._split_data(data)
+
+    def _split_data(self, data: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Split the raw data into a training set, and a testing(evaluation) set.
+        :param data: All the raw data
+        :return: traing, test raw data set
+        """
+        if self.test_mode:
+            self.test_set_size = len(data)
+        else:
+            self.test_set_size = self.eval_data_size
+
+        split_idx = len(data) - self.test_set_size
+        # First part as the training set
+        train_raw_data = data[:split_idx]
+
+        # Last part as the testing set
+        test_raw_data = data[split_idx:]
+
+        return train_raw_data, test_raw_data
+
+    @cached_property
+    def train_seqs(self) -> List[Tuple[np.ndarray, np.ndarray]]:
+        """
+        Create training sequences.  Each training sample consists of:
+         - A list of data points as input
+         - A label (target) value in some horizon
+        :return: training sequences
+        """
+        seq_len = self.seq_len
+        horizon = self.horizon_len
+        input_data = self.train_raw_data
+
+        seqs = []
+        for i in range(len(input_data) - seq_len - horizon):
+            seq = input_data[i:i + seq_len]
+
+            # Look beyond the horizon to get the label
+            label_i = i + seq_len + horizon
+            label = input_data[label_i: label_i + 1]
+
+            seqs.append((seq, label))
+        return seqs
+
+    @cached_property
+    def test_seqs(self) -> List[np.ndarray]:
+        """
+        Generate a list of test sequence for evaluation
+        :return:  Test sequences for evaluation
+        """
+        input_data = self.test_raw_data
+
+        seqs = []
+        for i in range(len(input_data) - self.seq_len):
+            seq = input_data[i:i + self.seq_len]
+            if self.x_transformer:
+                seq = self.x_transformer.transform(seq)
```

Comment:
This seems redundant? Furthermore, the transformation should be encapsulated inside the model anyway.